Lib/test/test_pkgutil.py

```diff
@@ -186,6 +186,46 @@ def test_walk_packages_raises_on_string_or_bytes_input(self):
         with self.assertRaises((TypeError, ValueError)):
             list(pkgutil.walk_packages(bytes_input))
 
+    def test_name_resolution(self):
+        import logging
+        import logging.handlers
+
+        cases = (
+            (None, TypeError),
+            (1, TypeError),
+            (2.0, TypeError),
+            (True, TypeError),
+            ('', ValueError),
+            ('?abc', ValueError),
+            ('abc/foo', ValueError),
+            ('foo', ImportError),
+            ('os', os),
+            ('os.path', os.path),
+            ('os.foo', AttributeError),
+            ('os.foo:', ImportError),
+            ('os.pth:pathsep', ImportError),
+            ('os.path:pathsep', os.path.pathsep),
+            ('logging', logging),
+            ('logging:', logging),
+            ('logging.handlers', logging.handlers),
+            ('logging.handlers:', logging.handlers),
+            ('logging.handlers:SysLogHandler', logging.handlers.SysLogHandler),
+            ('logging.handlers.SysLogHandler', logging.handlers.SysLogHandler),
+            ('logging.handlers:SysLogHandler.LOG_ALERT',
+                logging.handlers.SysLogHandler.LOG_ALERT),
+            ('logging.handlers.SysLogHandler.LOG_ALERT',
+                logging.handlers.SysLogHandler.LOG_ALERT),
+            ('logging.handlers:NoSuchHandler', AttributeError),
+            ('logging.handlers:SysLogHandler.NO_SUCH_VALUE', AttributeError),
+            ('logging.handlers.SysLogHandler.NO_SUCH_VALUE', AttributeError),
+        )
+
+        for s, e_or_o in cases:
+            try:
+                o = pkgutil.resolve_name(s)
+                self.assertEqual(o, e_or_o)
+            except Exception as e:
+                self.assertIsInstance(e, e_or_o)
```

Comment:
Success tests and error tests could be in two separate blocks:

```python
for s, expected in success_cases:
    with self.subTest(s=s):
        o = pkgutil.resolve_name(s)
        self.assertEqual(o, expected)

for s, exc in error_cases:
    with self.subTest(s=s):
        with self.assertRaises(exc):
            pkgutil.resolve_name(s)
```

I am a fan of making tests as simple as possible and avoid logic when possible, but that’s style so YMMV.

Reply:
That's fair, and I don't mind doing this. It does make things clearer.